src/_hyperscript.js

```diff
@@ -1244,6 +1244,31 @@
                     }
                 });
 
+                _parser.addGrammarElement("arrayIndex", function (parser, tokens, root) {
+
+                    if (tokens.matchOpToken("[")) {
+                        var indexExp = parser.parseElement("expression", tokens)
```

Comment:
I have introduced a new function `requireElement` that you probably want to use here, so it throws a parse error if no element is returned.